File: Chapter 04/Python/section_4_2_downward_dependency/before/some_repository.py
from .presentation_layer import PresentationLayer

class SomeRepository:
    """
    ARCHITECTURE WARNING: Upward Dependency Violation.
    """
    def __init__(self):
        # 🚨 VIOLATION: Dependency on an upper-layer singleton.
        # This couples our data logic to a specific UI implementation.
        self._ui_layer = PresentationLayer.get_instance()

    def update_data(self, user_id: int, new_data: str):
        print("(Before) Saving data to database...")
        
        # 🚨 VIOLATION: Calling upwards to the UI Layer.
        self._ui_layer.update_status_label(f"(Before) Data {user_id} Saved!")

File: Chapter 04/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter04/section_4_2_downward_dependency/before/SomeRepository.java
package com.grokkingsoftwarearchitecture.chapter04.section_4_2_downward_dependency.before;

/**
 * DataAccessLayer.java
 * * ARCHITECTURAL CRITIQUE:
 * This class represents an "Upward Dependency" violation. A lower 
 * layer (Infrastructure/Data Access) should never know about an 
 * upper layer (UI/Presentation).
 */
public class SomeRepository {
    // This is the violation! A lower layer should not
    // know about an upper layer.
    private PresentationLayer uiLayer = PresentationLayer.getInstance();

    public void updateData(int id, String newData) {
        // Normal print statement as requested
        System.out.println("(Before Refactoring) Saving data to database...");
        
        // VIOLATION! Calling upwards to the UI Layer.
        // This makes the repository impossible to test without 
        // bringing the entire UI with it.
        uiLayer.updateStatusLabel("(Before Refactoring) Data " + id + " Saved!");
    }
}